// telegram/entityCache.js
"use strict";
// Which updates have the following fields?
Object.defineProperty(exports, "__esModule", { value: true });
exports.EntityCache = void 0;
const Utils_1 = require("./Utils");
const Helpers_1 = require("./Helpers");
const tl_1 = require("./tl");
class EntityCache {
    constructor() {
        this.cacheMap = new Map();
    }
    add(entities) {
        const temp = [];
        if (!Helpers_1.isArrayLike(entities)) {
            if (entities != undefined) {
                if (typeof entities == "object") {
                    if ("chats" in entities) {
                        temp.push(...entities.chats);
                    }
                    if ("users" in entities) {
                        temp.push(...entities.users);
                    }
                    if ("user" in entities) {
                        temp.push(entities.user);
                    }
                }
            }
            if (temp.length) {
                entities = temp;
            }
            else {
                return;
            }
        }
        for (const entity of entities) {
            try {
                const pid = Utils_1.getPeerId(entity);
                if (!this.cacheMap.has(pid)) {
                    this.cacheMap.set(pid, Utils_1.getInputPeer(entity));
                }
            }
            catch (e) { }
        }
    }
    get(item) {
        if (!(typeof item === "number") || item < 0) {
            let res;
            try {
                res = this.cacheMap.get(Utils_1.getPeerId(item));
                if (res) {
                    return res;
                }
            }
            catch (e) {
                throw new Error("Invalid key will not have entity");
            }
        }
        for (const cls of [tl_1.Api.PeerUser, tl_1.Api.PeerChat, tl_1.Api.PeerChannel]) {
            // TODO remove these "as"
            const result = this.cacheMap.get(Utils_1.getPeerId(new cls({
                userId: item,
                chatId: item,
                channelId: item,
            })));
            if (result) {
                return result;
            }
        }
        throw new Error("No cached entity for the given key");
    }
}
exports.EntityCache = EntityCache;
